### ISSUE TITLE
python destination, imports() option

### ISSUE THREAD

**Issue by mitzkia** [balabit/syslog-ng#1361]
* If I configure python destination with imports() option, I can not use imported module from python script, however the import worked (see the "sys.modules" result in example)
 * Importing new modules works correctly if I imported them from the python script (see "import sys" in example below)
 * I think this option (imports()) is unnecessary, the specific module also can be imported from the used python script.

Reproduction:

syslog-ng.conf
```
@version: 3.9
source s_net {tcp( ip("127.0.0.1") port(6666) );};
destination d_python {
  python(
    class("DummyPythonDest")
    imports('time')
  );
};
log {
  source(s_net);
  destination(d_python);
};
python {
class DummyPythonDest(object):
    def init(self, options):
        #import time
        import sys
        print(sys.modules)
        time.sleep(1)
        return True
    def send(self):
        return True
};
```

console log:
```
{'google': <module 'google' (built-in)>, 'copy_reg': <module 'copy_reg' from '/usr/lib/python2.7/copy_reg.pyc'>, 'sre_compile': <module 'sre_compile' from '/usr/lib/python2.7/sre_compile.pyc'>, '_sre': <module '_sre' (built-in)>, 'encodings': <module 'encodings' from '/usr/lib/python2.7/encodings/__init__.pyc'>, 'site': <module 'site' from '/usr/lib/python2.7/site.pyc'>, '__builtin__': <module '__builtin__' (built-in)>, 'zc': <module 'zc' (built-in)>, 'sysconfig': <module 'sysconfig' from '/usr/lib/python2.7/sysconfig.pyc'>, '__main__': <module '__main__' (built-in)>, 'encodings.encodings': None, 'abc': <module 'abc' from '/usr/lib/python2.7/abc.pyc'>, 'posixpath': <module 'posixpath' from '/usr/lib/python2.7/posixpath.pyc'>, '_weakrefset': <module '_weakrefset' from '/usr/lib/python2.7/_weakrefset.pyc'>, 'errno': <module 'errno' (built-in)>, 'encodings.codecs': None, 'sre_constants': <module 'sre_constants' from '/usr/lib/python2.7/sre_constants.pyc'>, 're': <module 're' from '/usr/lib/python2.7/re.pyc'>, '_abcoll': <module '_abcoll' from '/usr/lib/python2.7/_abcoll.pyc'>, 'types': <module 'types' from '/usr/lib/python2.7/types.pyc'>, '_codecs': <module '_codecs' (built-in)>, 'encodings.__builtin__': None, '_warnings': <module '_warnings' (built-in)>, 'genericpath': <module 'genericpath' from '/usr/lib/python2.7/genericpath.pyc'>, 'stat': <module 'stat' from '/usr/lib/python2.7/stat.pyc'>, 'zipimport': <module 'zipimport' (built-in)>, '_sysconfigdata': <module '_sysconfigdata' from '/usr/lib/python2.7/_sysconfigdata.pyc'>, 'warnings': <module 'warnings' from '/usr/lib/python2.7/warnings.pyc'>, 'UserDict': <module 'UserDict' from '/usr/lib/python2.7/UserDict.pyc'>, 'encodings.utf_8': <module 'encodings.utf_8' from '/usr/lib/python2.7/encodings/utf_8.pyc'>, 'sys': <module 'sys' (built-in)>, 'codecs': <module 'codecs' from '/usr/lib/python2.7/codecs.pyc'>, '_sysconfigdata_nd': <module '_sysconfigdata_nd' from '/usr/lib/python2.7/plat-x86_64-linux-gnu/_sysconfigdata_nd.pyc'>, 'os.path': <module 'posixpath' from '/usr/lib/python2.7/posixpath.pyc'>, '_syslogng': <module '_syslogng' (built-in)>, '_locale': <module '_locale' (built-in)>, 'sitecustomize': <module 'sitecustomize' from '/usr/lib/python2.7/sitecustomize.pyc'>, 'signal': <module 'signal' (built-in)>, 'traceback': <module 'traceback' from '/usr/lib/python2.7/traceback.pyc'>, 'apport_python_hook': <module 'apport_python_hook' from '/usr/lib/python2.7/dist-packages/apport_python_hook.pyc'>, 'linecache': <module 'linecache' from '/usr/lib/python2.7/linecache.pyc'>, 'posix': <module 'posix' (built-in)>, 'encodings.aliases': <module 'encodings.aliases' from '/usr/lib/python2.7/encodings/aliases.pyc'>, 'time': <module 'time' (built-in)>, 'exceptions': <module 'exceptions' (built-in)>, 'sre_parse': <module 'sre_parse' from '/usr/lib/python2.7/sre_parse.pyc'>, 'os': <module 'os' from '/usr/lib/python2.7/os.pyc'>, '_weakref': <module '_weakref' (built-in)>}
Traceback (most recent call last):
  File "<string>", line 7, in init
NameError: global name 'time' is not defined

```

**Comment by furiel** [comment]
The interesting thing is the same behaviour occurs with the pure python interpreter, without syslog-ng involved. Just create a test.py file:
```
def testfunc():
    time.sleep(1)
```
Setting `export PYTHONPATH=.` and starting the interpreter:
```
>>> import test
>>> import time # <--
>>> test.testfunc()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "test.py", line 2, in testfunc
    time.sleep(1)
NameError: global name 'time' is not defined
```
So the imports option is definitely not intended to use for eliminating imports from the user provided code. I wonder what would be the usecase.

Unfortunately I did not find any reference in the administration guide, neither in the development guide in the gitbook. Tracking back the origin: this option originates from the incubator,  in 
https://github.com/balabit/syslog-ng-incubator/commit/fe61f0b49c5601a3cdf40d1d737081c72d1348ff
```
This works in a very similar way to the Perl destination, except that it
is for python, and supports an imports() argument too, which can list
modules to pre-load before the script.
```

@algernon if you happen to remember to this old patch, can you explain how this imports option to be used, and what was the use case behind it?

**Comment by bazsi** [comment]
The use case was to import the hy module, that registers import hooks to
translate hy source code to python bytecode.

**Comment by furiel** [comment]
Thanks for explaining, definitely got my attention! Very inspiring to know such feature is available!

It would worth writing some short description and use case about this in the documentation, so that we can avoid further confusion on this option.

After a few trial and error I could fire up the hy module. Though I ran into a few **caveats**.
```
mymodule.hy:
(defclass MyPyDest [object]
  (defn init [self options]
    True)
  (defn send [self message]
    (print (get message (bytes "MESSAGE")))
    True))

syslog-ng.conf
destination { python(imports("hy") class("mymodule.MyPyDest")); };
```

First `imports("hy")` failed with
```Error loading Python module; module='hy', exception='exceptions.ImportError: cannot import name argv'```

The reason is the hy module imports the clint module, which in terms calls  [sys import argv](https://github.com/kennethreitz/clint/blob/9d3693d644b8587d985972b6075d970096f6439e/clint/arguments.py#L15), which is a problem, because in c binding sys.argv is not set. I could make this work with a local patch in clint:
```
if hasattr(sys, 'argv'):
    from sys import argv
```

The next problem arrived when I tried `(get message "MESSAGE")`. This led to
```Exception while calling a Python function; module='#anon-destination0#0', script='mymodule.MyPyDest', function='send', exception='exceptions.SystemError: error return without exception set'```
It turns out the hook _py_log_message_getattr in syslog-ng returns with nullpointer in case of error. Instead of this [Py_RETURN_NONE](https://docs.python.org/2/c-api/none.html) should be written.
Now the system error disappeared, but I got None instead of the actual message. It turned out the code that checks if the argument is string in `_py_log_message_getattr` caused this failure. 
Here there are two issues, there is no any feedback to the user if not a string is passed to the message, although currently it is a requirement. Here syslog-ng could throw a `TypeError`.
To get it working, I needed to cast `"MESSAGE"` to bytes using `(print (get message (bytes "MESSAGE")))`, which is an unnecessary complexity on the user interface. Probably this behaviour was intrduced by https://github.com/balabit/syslog-ng/pull/1632, though the whole case is interesting, because without the hy module, one can simply write `print msg["MSG"]`, and it works as expected.

In summary, I think the the following points need to be resolved in this ticket.
- Documentation: need to add a description, usecase and example code, explaining how this option can be utilized.
- Fix the import error for hy module
- _py_log_message_getattr, _py_log_message_setattr (maybe others) should return with None object instead of NULL.
- Adding exception when wrong type is used as key in message[key].
- Check if the bytes cast can be eliminated in `(print (get message (bytes "MESSAGE")))`
- Add an e2e testcase for this option.
